From b927a6c9a3791febce672e62f092ff00bae9546c Mon Sep 17 00:00:00 2001
From: delthas <delthas@dille.cc>
Date: Sun, 16 Oct 2022 03:12:30 +0200
Subject: [PATCH] crt: Open the mkstemp temporary file without denying RW
 access

Previously, mkstemp opened its temporary file as _SH_DENYRW, which
explicitly prevented any other access to this file from another fd while
this fd was still open.

This is because we passed the parameter _SH_DENYRW to _sopen, which
according to the Windows documentation: "Denies read and write access to
a file."

However, mkstemp is not supposed to prevent other concurrent access to
that file. The Linux man page for open(2) specifies no such behavior.
Further, the glibc current implementation is equivalent to:
  open(template_name, O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
which does not involve preventing other access to this file.

To get the same behavior as glibc and the Linux man page, we should
instead allow other access to that temporary file, by passing the
_SH_DENYNO parameter to _sopen, which according to the Windows
documentation: "Permits read and write access."

This patch fixes this behavior by updating the shared access parameter
from _SH_DENYRW to _SH_DENYNO.

As an example of why this could be harmful, consider the following
excerpt. This is arguably unoptimal code, but comes from an actual
program used in production.

  char temp = "tmpXXXXXX";
  int fd1 = mkstemp(temp);
  // ...
  int fd2 = fopen(temp, "wb");
  // ... use fd2
  fclose(fd2);
  // ...
  unlink(fd1);

This program uses mkstemp to get a name of a temporary file, but does
not use the fd it gets from opening the file. mkstemp is merely used to
generate a temporary file name.

On glibc, this would work as expected.

On MinGW, this would fail at the fopen step, with a return code of -1
and an errno set to Permission Denied, because the mkstemp opened fd1
with _SH_DENYRW which prevents any other access to the file.

After this commit, the MinGW implementation will work as expected.

Signed-off-by: delthas <delthas@dille.cc>
Signed-off-by: LIU Hao <lh_mouse@126.com>
---
 mingw-w64-crt/misc/mkstemp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mingw-w64-crt/misc/mkstemp.c b/mingw-w64-crt/misc/mkstemp.c
index c78b7c509..6b327f2fc 100644
--- a/mingw-w64-crt/misc/mkstemp.c
+++ b/mingw-w64-crt/misc/mkstemp.c
@@ -49,7 +49,7 @@ int __cdecl mkstemp (char *template_name)
         }
         fd = _sopen(template_name,
                 _O_RDWR | _O_CREAT | _O_EXCL | _O_BINARY,
-                _SH_DENYRW, _S_IREAD | _S_IWRITE);
+                _SH_DENYNO, _S_IREAD | _S_IWRITE);
         if (fd != -1) return fd;
         if (fd == -1 && errno != EEXIST) return -1;
     }
